opentracing/span.go

```diff
@@ -0,0 +1,165 @@
+package opentracing
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	opentracing "github.com/opentracing/opentracing-go"
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// Span is the Appdash implementation of the `opentracing.Span` interface.
+type Span struct {
+	sync.Mutex
+	Recorder      *appdash.Recorder
+	tracer        *Tracer
+	operationName string
+	startTime     time.Time
+	sampled       bool
+	attributes    map[string]string
+	tags          map[string]interface{}
+
+	logs []opentracing.LogData
+}
+
+func newAppdashSpan(operationName string, tracer *Tracer) *Span {
+	return &Span{
+		operationName: operationName,
+		tracer:        tracer,
+		startTime:     time.Now(),
+		logs:          make([]opentracing.LogData, 0),
+	}
+}
+
+// SetOperationName sets the name of the span, overwriting the previous value.
+func (s *Span) SetOperationName(operationName string) opentracing.Span {
+	s.operationName = operationName
+	return s
+}
+
+// Tracer returns the interal opentracing.Tracer
+func (s *Span) Tracer() opentracing.Tracer {
+	return s.tracer
+}
+
+// Finish ends the span.
+//
+// Defers to s.FinishWithOptions() with FinishTime = time.Now()
+func (s *Span) Finish() {
+	s.FinishWithOptions(opentracing.FinishOptions{FinishTime: time.Now()})
+}
+
+// FinishWithOptions finishes the span with opentracing.FinishOptions.
+//
+// Internally, the `appdash.Reporter` reports the span's name, tags,
+// attributes, and log events.
+func (s *Span) FinishWithOptions(opts opentracing.FinishOptions) {
+	if !s.sampled {
+		return
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	s.Recorder.Name(s.operationName) // Set the span's name
+
+	// Convert span tags to annotations.
+	for key, value := range s.tags {
+		val := []byte(fmt.Sprintf("%v", value))
+		s.Recorder.Annotation(appdash.Annotation{Key: key, Value: val})
+	}
```

Comment:
Might consider using a custom event struct for transmitting the `s.tags` payload itself. Actually, I'm not too sure. But I think `key` might conflict with how we unmarshal annotations into structures.. on the other hand, it might not and using an event struct adds an overhead.. hmm. I'll look into this and get back to you.

Did you test this in the frontend UX after adding a tag to a span ? did it render the trace OK and the tag k/v pair showed up in the "Verbose Data View" tab of the trace view?